In [1]:
import pandas as pd

df = pd.read_csv("messy_sales.csv")
print(df.shape)
print(df.head(10).to_string(index=False))
print(df.dtypes)

(300, 6)
 order_id       date  product  price  qty   zip
     1254 04/06/2026      mug  11.36    4 60614
     1057 2026-05-30  charger  14.79    3 30303
     1150 05/06/2026      mug   5.50    1 10001
     1066 05/30/2026 notebook    NaN    1 98101
     1114 04/06/2026   webcam  45.59    4 90405
     1280 04/22/2026  pen set    NaN    4 30303
     1204 04/18/2026  charger   8.41    2  2134
     1249 2026-05-15 keyboard  75.04    4 90405
     1037 2026-05-19      mug   9.31    2 60614
     1177 2026-05-13      mug  10.11    3 90210
order_id      int64
date         object
product      object
price       float64
qty           int64
zip           int64


1. Certain values are not there in the price field.

2. The ZIP codes have been entered in numeric form which made certain leading zeroes disappearing.

3. The dates have been entered in varying formats and will have to be converted to one format.

In [2]:
print(df.isna().sum())

order_id     0
date         0
product      0
price       12
qty          0
zip          0


In [3]:
fill_value = df["price"].median()
df["price"] = df["price"].fillna(fill_value)
print(fill_value)
print(df.isna().sum())

37.53
order_id    0
date        0
product     0
price       0
qty         0
zip         0


In [4]:
print(df.duplicated().sum())
df = df.drop_duplicates()
print(df.shape)

8
(292, 6)


In [5]:
df["zip"] = df["zip"].astype(str).str.zfill(5)
print(df["zip"].head(10).to_string(index=False))

60614
30303
10001
98101
90405
30303
02134
90405
60614
90210


In [6]:
df["date"] = pd.to_datetime(df["date"], format="mixed")
print(df.dtypes)

order_id             int64
date        datetime64[ns]
product             object
price              float64
qty                  int64
zip                 object


In [7]:
print(df.shape)
print(df["zip"].str.len().eq(5).all())
print(df["date"].dtype)

(292, 6)
True
datetime64[ns]


In [8]:
return_mask = df["qty"] < 0
print(df.loc[return_mask].to_string(index=False))
print("Negative quantities retained:", return_mask.sum())

 order_id       date   product  price  qty   zip
     1140 2026-04-18    webcam  38.69   -5 98101
     1233 2026-05-09 desk lamp  22.64   -4 10001
     1025 2026-04-27  keyboard  47.30   -2 02116
Negative quantities retained: 3


There were three entries where the values were in negative figures. I opted to retain such entries as they could be representing customer’s refunds rather than errors in the data. Changing them may lead to incorrect sales data.

In [9]:
df.to_csv("sales_clean.csv", index=False, date_format="%Y-%m-%d")
print(df.shape)
print("sales_clean.csv saved")

(292, 6)
sales_clean.csv saved


## Cleaning Log

The database contained 300 rows. Missing prices were found in 12 rows, but I decided to use the median of $37.53 to fill the blanks since the median price is more robust compared to outliers. I found and deleted 8 duplicated rows, and now there are 292 rows remaining.

I changed ZIP codes to string data type and increased its length up to five digits to keep leading zeros. The date was converted to date time data type.

Three rows contained negative quantities, but I did not delete these rows because they might show returned products, rather than errors.

Bad choice in cleaning process will make a company receive inaccurate results from sales.

Repo: https://github.com/washingtonjustin02smc/CS82A-repository

## AI Use Appendix

Tool: ChatGPT

Key prompts: Help clean messy_sales.csv, verify the pandas results, create the notebook, cleaned CSV, and PDF, and update the notebook with my rephrased answers.